### ISSUE TITLE
Nesting leads to multiple mounts

### ISSUE THREAD

**Issue by RohovDmytro** [expo/router#733]
### Which package manager are you using? (Yarn is recommended)

yarn

### Summary

A quite simple navigation structure leads to screens beings mounted multiple times.

Memoization does not help. the screen remounts, not re-renders

I would argue that this bug is close to critical as many things are tight to component being mounted. Such features as analytics / data loading / animations become sadly broken.

I've spent quite some time restructuring components trying to avoid nesting, but in a more complicated scenarios (with auth, screen groups, etc) it's very hard. Found zero workarounds.

### Minimal reproducible example

https://github.com/RohovDmytro/expo-router-has-a-fundamental-issue

**Comment by EvanBacon** [comment]
Working on the issue now, it appears to only happen on native platforms, when the root layout is a `<Stack />` and it has a nested layout. This can be worked around by making the root a `<Slot />` and moving everything up a layer using groups.

```
_layout -> <Slot />
  (app)
    _layout -> <Stack />
```

Given the current symptoms, it would seem like this might be a `react-native-screens` issue.

**Comment by RohovDmytro** [comment]
My setup leads to 30 FPS drop per screen transition (Google Pixel 7) between certain screens if Slot is in the root. First impression is that navigating between Groups is very, very expensive in terms of performance (when root is Slot).

It's 150 FPS dropped (with Slot as root) vs 30 FPS dropped (with multiple re-renders) while onboarding a user.

I will keep looking for a way to restructure routes and will make a notice.

**Comment by EvanBacon** [comment]
So far I've traced this to upstream so it'll take a while for a proper fix. I can't reproduce any "very, very expensive" drop in performance by introducing an additional empty navigator.

**Comment by RohovDmytro** [comment]
> So far I've traced this to upstream so it'll take a while for a proper fix. I can't reproduce any "very, very expensive" drop in performance by introducing an additional empty navigator.

Thanks, @EvanBacon, for tracing and spending your time on this issue! I appreciate. 

Spent lots of time workarounding the 0.71 react-native perf regression, so having high hopes of efficiently rendering components during the second part of the year  :)

Yeah. Let me share the details. This is what I'm using to get a 30 FPS drop (Google Pixel 7, dev enabled):

- The root layout is using a `Slot`. 
- The nested navigation is a `Stack`. 
- If a screen component placed in a nested navigator is wrapped in a `SafeAreaView` (from `react-native-safe-area-context`) there's when big FPS drop happening.

Not reporting this as `expo-router` bug, rather than sharing the current downside of the workaround. Not sure the relation between expo-router and safe-areas-component yet.

**Comment by RohovDmytro** [comment]
One last bits of info before snoozing myself:

- Adding `SafeAreaProvider` to the root `_layout` makes no difference, transition is still lagging.

- Using `SafeAreView` from `react-native-safe-area-view` works surprisingly well, no dropped frames. The problems is `react-native-safe-area-view` is clearly deprecated and the docs are saying to stay away from it.

````
⚠️ This library is deprecated. It is no longer used in React Navigation and it has been succeeded by the excellent [react-native-safe-area-context](https://github.com/th3rdwave/react-native-safe-area-context). Please use react-native-safe-area-context instead, or you are likely to have a bad time.
````

**Comment by RohovDmytro** [comment]
I've updated the repo with a case when the bug persists after restructuring layouts:

https://github.com/RohovDmytro/expo-router-has-a-fundamental-issue/commit/661b7bbe5420502891f4869584e12cce3fa85ae4

The workaround works in a simplest case, but introducing route organization (groups) does break things.

Using the terminology from the commit above:

Slot => Slot1 => ✅
Slot1 => Slot2 => ❌ (unnecessary mounts). Slot1 (the starting screen) is mounted again while navigating to Slo2.

(If keep moving between screen, e.g. navigating Slot2 => Slot1 works fine, but then navigating Slot1 => Slot2 lead to mounting Slot2 screen twice).

**Comment by nicramu** [comment]
I'm facing same issue. What I did is, I deleted the _layout file in subfolder and defined the Stack Screens in parent _layout file. So the structure is as follows:
```
>app
_layout
index
  >subfolder
  [site]
  ```
  
  and I'm adding screens in _layout as follows:
  ```
   const {title} = useSearchParams()
   
        <Stack initialRouteName="index" screenOptions={{ headerShown: false, statusBarTranslucent: true, }}>
          <Stack.Screen name="index" options={{ headerShown: false }} />
          <Stack.Screen name="subfolder/[site]" options={{ headerShown: true, title: title, }} />
        </Stack>
  ```

as you can see even useSearchParams is used and I'm not getting any double execution of code. It may be that not having _layout file in subfolder is cutting some features? but I'm not aware of anything like that and this workaround seems to work for me for now. Worth giving it a try in your case also.
![scr](https://github.com/expo/router/assets/110261550/2ab2bc9f-c281-4d1a-a481-564c9b70c211)

**Comment by olarra** [comment]
I confirm this fix prevent nested routes to be re-rendered

**Comment by EvanBacon** [comment]
A fix has been published in `expo-router@2.0.3` (`2.0.3`). Sorry for any inconvenience this may have caused.
